include/revng-c/mlir/Dialect/Clift/IR/CliftOps.td

```diff
@@ -26,6 +27,120 @@ def Clift_StatementRegion : Region<CPred<"::mlir::clift::impl::verifyRegionConte
 def Clift_ExpressionRegion : Region<CPred<"::mlir::clift::impl::verifyRegionContent<::mlir::clift::ExpressionOpInterface>($_self)">,
                                     "Region representing an expression">;
 
+//===------------------------ Top level operations ------------------------===//
+
+def Clift_ModuleOp : Clift_Op<"module",
+                              [SymbolTable,
+                               HasOnlyGraphRegion,
+                               NoRegionArguments,
+                               NoTerminator,
+                               SingleBlock,
+                               RegionKindInterface]> {
+  let regions = (region AnyRegion:$body);
+
+  let skipDefaultBuilders = 1;
+  let builders = [OpBuilder<(ins)>];
+
+  let assemblyFormat = [{
+    $body attr-dict
+  }];
+
+  let hasVerifier = 1;
+}
+
+def Clift_FunctionOp : Clift_Op<"func",
+                                [Symbol,
+                                 CallableOpInterface,
+                                 FunctionOpInterface,
+                                 AutomaticAllocationScope,
+                                 IsolatedFromAbove,
+                                 HasParent<"mlir::clift::ModuleOp">]> {
+  let arguments = (ins StrAttr:$sym_name,
+                       TypeAttrOf<FunctionType>:$function_type,
+                       OptionalAttr<DictArrayAttr>:$arg_attrs,
+                       OptionalAttr<DictArrayAttr>:$res_attrs);
```

Comment:
This is different from the spec. Arguments and return types should be entirely implied by the `FunctionType`.

Reply:
See here: https://github.com/revng/llvm-project/blob/master/mlir/include/mlir/IR/FunctionInterfaces.td#L37

Reply:
Looking at how it's used in the interface though, maybe we can actually replace these arguments with functions.

Reply:
Ok this is done now.